src/cmap/connection_pool.ts

```diff
@@ -572,6 +573,66 @@ export class ConnectionPool extends TypedEventEmitter<ConnectionPoolEvents> {
     });
   }
 
+  private withReauthentication(
+    fnErr: AnyError,
+    conn: Connection,
+    fn: WithConnectionCallback,
+    callback: Callback<Connection>
+  ) {
+    if ((fnErr as MongoError).code === MONGODB_ERROR_CODES.Reauthenticate) {
+      this.reauthenticate(conn, fn, (error, res) => {
+        if (error) {
+          return callback(error);
+        }
+        callback(undefined, res);
+      });
+    } else {
+      callback(fnErr);
+    }
+  }
+
+  /**
+   * Reauthenticate on the same connection and then retry the operation.
+   */
+  private reauthenticate(
+    connection: Connection,
+    fn: WithConnectionCallback,
+    callback: Callback
+  ): void {
+    const authContext = connection.authContext;
+    if (!authContext) {
+      return callback(new MongoRuntimeError('No auth context found on connection.'));
```

Comment:
There's a number of synchronous callback invocations, are we sure that's the correct approach? I could see us implementing this using async/await and wrapping the old APIs in promisify.

- reauth could promisify auth, and use try/finally to set/unset the 
- we can promisify calling the `fn` 
- `withReauthentication` can return a promise and we can invoke the withConnection callback in a `.then`

thoughts?

Reply:
I think that's a good idea but maybe we can put that in a separate refactoring ticket? When I first went down that path I started leaking connections and felt the refactor was getting a bit more out of scope from just this ticket.

Reply:
I am targeting all the new code here with the suggested changes, just trying to make all new stuff async await and we can use .then & etc to put it inside of callback existing code. However, I don't feel super strongly so a follow up sounds good

Reply:
I think in the future this is something that we should try to capture in grooming or at least in the kickoff notes - the intent to use async/await or refactor - to help set the scope of the work and align with the team early on. As is, and considering that right now we want to have these changes in place and passing the new prose tests, a follow up ticket seems like the most viable approach - though I'm not certain where that work would live. In general, we do want to try to account for the async/await refactor when estimating new tickets because it really helps with the async stack traces (which was the original motivation for moving the driver to async/await).

Reply:
I generally try to async/await everything when introducing new code, and refactoring the surrounding code as well. (See recent PRs in this area). This particular area was a little bit more risky to me in this PR as everything in the driver runs through this path - once some leaks started happening I aborted. I think maybe this can go in a ticket/epic around refactoring the connection layer?

Reply:
yup, sounds good to me